### ISSUE TITLE
Rendering artifacts in navigation mode when perspective view is on.

### ISSUE THREAD

**Issue by Endem10n** [organicmaps/organicmaps#891]
After created route to some point, some objects rendered with artifacts, when perspective view setting is on.

Coordinates, where it can be reproduced: 54°24'50.0"N 18°36'12.6"E; 54°24'27.4"N 18°36'59.7"E.

Reproduced on Vivo Y11s with Android 11 and on iPhone SE with iOS 14.6

[video from Android](https://drive.google.com/file/d/1JdBfb2FWQnmoYonv10TqEdnwXLHSg3Jr/view?usp=sharing)
[video from iOS](https://drive.google.com/file/d/18ujuA0-UL7KiO6KJv30CFZ0fizzbDWkO/view?usp=sharing)

**Comment by vng** [comment]
Reproduced flickering here 49.4219568,8.6382218 on desktop, perspective:
<img width="883" alt="Screenshot 2022-04-12 at 14 42 04" src="https://user-images.githubusercontent.com/175612/162952620-2b4031d8-72b4-44e2-9344-e6e60cdb955b.png">

**Comment by vng** [comment]
- ALT + LMB to set your current position
- SHIFT + LMB to set end route point and build route
- Press blue arrow button (Follow route)

Emulate driving positions with ALT + LMB

**Comment by pastk** [comment]
> Reproduced flickering here 49.4219568,8.6382218 on desktop, perspective: 

Now OSM data has changed here, but it looks like it was `landuse-allotments` mapped on top of `landuse-farmland` - their area drules have the same priority and hence polygons sent to OGL had the same depth value. So the flickering is probably a result of "z-fighting".

Since #4581 it shouldn't happen in such cases as the depth value is derived from area's bbox sizes and here they're very different.

But now it could happen for cases when same bbox size features overlap each other, like here: https://www.openstreetmap.org/way/616385905 (a cemetery and a forest mapped on top of each other). As bbox sizes are the same then depth values are the same and hence "z-fighting" resulting in flickering:
[video](https://drive.google.com/file/d/1MyhbkU3gb5TrYC_QHUI8GcWVtBUvnQpu/view)

This is the hypothesis :)
If its true then to prevent "z-fighting" we need to avoid identical depth values - we can try to add a priority-derived tiny value to bbox size value - so tiny as to not affect the size ordering, but big enough to make depth values different (i.e. distinguishable given a limited z-buffer precision).

**Comment by RicoElectrico** [comment]
This issue also happens for me in the latest beta:
https://www.openstreetmap.org/#map=17/54.47004/18.52403

https://github.com/organicmaps/organicmaps/assets/1409116/8d1b9166-6a41-4bff-aa99-78f1dc9b837c

**Comment by biodranik** [comment]
Is it an OSM data issue that should be fixed there?

**Comment by RicoElectrico** [comment]
@biodranik Yes, I fixed it yesterday. However I think it's best to handle errors gracefully.
I know it's not that rare for mappers to come up with such constructs. For example when cemetery sectors did not have standardized tagging, people mapped `landuse=cemetery`+`name=*` inside `landuse=cemetery`. Likewise I would not be surprised if I see `landuse=residential` inside `landuse=residential` to signify housing estates named by the property developers.
@pastk maybe we can do something similar to openstreetmap-carto which sorts z-order by area?

**Comment by pastk** [comment]
@RicoElectrico we sort by area size already (bounding box though, not a true precise area), but in those cases bbox areas are of same size, hence the issue. A possible solution I described before will work for overlapping features of different types. For the same type (duplicate) features we might add a tiny random or featureId derived value.
If you're curious or want help fix it, then the relevant code is in `drape_frontend/stylist.cpp Aggregator::Init()`.

**Comment by RicoElectrico** [comment]
Oh, I see. This error indeed will be very rare as it not only requires geometry to cross, but also bboxes to equal.
Here's an example PostGIS query:
`select a.osm_id, b.osm_id, a.landuse from planet_osm_polygon a join planet_osm_polygon b on a.landuse is not null and a.landuse=b.landuse and a.osm_id < b.osm_id and a.way && b.way and ST_Relate(a.way,b.way,'2********') and ST_Envelope(a.way) = ST_Envelope(b.way) limit 1;`
which returns such cases for landuse.
I don't know why https://www.openstreetmap.org/way/243870368 and https://www.openstreetmap.org/relation/10755784 don't flicker, then?

**Comment by pastk** [comment]
> I don't know why https://www.openstreetmap.org/way/243870368 and https://www.openstreetmap.org/relation/10755784 don't flicker, then?

Those ways are a little different (you've got to take a very close look), so their bbox sizes might be a little different too. I don't know how your query works, perhaps the query's and OM's precision of calculation is different. You might want to log the sizes in OM to confirm.

**Comment by pastk** [comment]
> I don't know why https://www.openstreetmap.org/way/243870368 and https://www.openstreetmap.org/relation/10755784 don't flicker, then?

Those ways are a little different (you've got to take a very close look), so their bbox sizes might be a little different too. I don't know how your query works, perhaps the query's and OM's precision of calculation is different. You might want to log the sizes in OM to confirm.

**Comment by RicoElectrico** [comment]
> I don't know how your query works, 
Let me explain :)
`~=` is an operator that, using an index, compares whether 2 geometries have an identical bbox
ST_Relate does a generic spatial relationship tester, with 9 elements that are combination of interior/boundary/exterior of 2 geometries. Here I'm testing whether interior of the intersection of 2 geometries has a dimensionality of 2 (that is, it has non-zero area).

**Comment by biodranik** [comment]
If it's OSM data issues, then IMO we should not waste time on them, but instead fix them, and let everyone know: see blinking? Fix it.

**Comment by 0Karakurt0** [comment]
50°26′58.27″N, 30°30′09.3″E

Reproduced here

**Comment by fizwidget** [comment]
Reproduced [here](https://www.openstreetmap.org/#map=16/-33.8496/151.2460&layers=N) (in pedestrian navigation mode). 

![IMG_3611](https://github.com/organicmaps/organicmaps/assets/1172012/dbd3279b-a2cb-4842-ad27-caf0d89a7af0)

**Comment by pastk** [comment]
In this case it seems to be [`natural=wood`](https://www.openstreetmap.org/relation/5900272) vs [`leisure=park`](https://www.openstreetmap.org/relation/2363738) area fills that seemingly share same polygon borders.

**Comment by user334** [comment]
Should we report here all the places where this issue occurs? 🤔
Anyway, found it happening here: [https://omaps.app/44Ietr4urv/Лес](https://omaps.app/44Ietr4urv/Лес)
App version: `2024.07.08-2-Google-beta` with the latest available maps. 
Video: https://github.com/user-attachments/assets/9948d4fc-ffb4-474b-99b7-9ba7550c2d75

**Comment by user334** [comment]
Reproduced here: https://omaps.app/w4I0DjqIhK/Uilenbos 
There is a military zone which is shown on maps via red stroked overlay. This exact overlay flickers there while navigating.

**Comment by dmitrygribenchuk** [comment]
Hello. Have the same since 2023.09.05-12-Google-beta

Coordinates: https://www.openstreetmap.org/#map=19/54.645681/25.282650

Uploading Screenrecorder-2023-09-07-20-16-32-511_compressed.mp4…

**Comment by dmitrygribenchuk** [comment]
Hello. Have the same since 2023.09.05-12-Google-beta

Coordinates: https://www.openstreetmap.org/#map=19/54.645681/25.282650

https://github.com/user-attachments/assets/a78db3f7-d735-4d34-a5f3-81efd62892a6

**Comment by Self-Perfection** [comment]
I get the same issue here https://www.openstreetmap.org/#map=19/48.865820/2.317627

This is how it looks like:

https://github.com/user-attachments/assets/462bde7a-9336-4ba4-937f-a0cc9e246f57

I suspect these two relations are in conflict:
* https://www.openstreetmap.org/relation/1060283/
* https://www.openstreetmap.org/relation/10142348

Does not look like incorrect mapping to me. But I am not quite proficient with relations.

**Comment by Felix14-v2** [comment]
I have similar issue here: https://omaps.app/04frygyIO_/

https://github.com/user-attachments/assets/696df192-f460-49a4-bd72-2303020e9972

geo:56.1437838,40.4352441?z=17.0&q=56.1437838,40.4352441(%D0%9B%D0%B5%D1%81)
